src/script/calling/CallingRepository.ts

```diff
@@ -683,16 +685,39 @@ export class CallingRepository {
   // Call actions
   //##############################################################################
 
-  private readonly toggleState = (withVideo: boolean): void => {
+  private readonly toggleState = async (withVideo: boolean): Promise<void> => {
     const conversation = this.conversationState.activeConversation();
     if (conversation) {
       const isActiveCall = this.findCall(conversation.qualifiedId);
       const callType = withVideo ? CALL_TYPE.VIDEO : CALL_TYPE.NORMAL;
-      return (
-        (isActiveCall
-          ? this.leaveCall(conversation.qualifiedId, LEAVE_CALL_REASON.ELECTRON_TRAY_MENU_MESSAGE)
-          : this.startCall(conversation, callType)) && undefined
-      );
+
+      if (isActiveCall) {
+        this.leaveCall(conversation.qualifiedId, LEAVE_CALL_REASON.ELECTRON_TRAY_MENU_MESSAGE);
+        return;
+      }
+
+      const memberCount = conversation.participating_user_ets().length;
+
+      if (memberCount > MAX_USERS_TO_CALL_WITHOUT_CONFIRM) {
+        PrimaryModal.show(PrimaryModal.type.WITHOUT_TITLE, {
+          preventClose: true,
+          primaryAction: {
+            action: async () => await this.startCall(conversation, callType),
+            text: t('groupCallModalPrimaryBtnName'),
+          },
+          secondaryAction: {
+            text: t('modalConfirmSecondary'),
+          },
+          text: {
+            htmlMessage: `<div class="modal-description">
+            ${t('groupCallConfirmationModalTitle', memberCount)}
+          </div>`,
+            closeBtnLabel: t('groupCallModalCloseBtnLabel'),
+          },
+        });
+      } else {
```

Comment:
This code must exist somewhere else. We should make sure that the calling button AND the desktop shortcut both go through the exact same code!

Also modals triggering should belong in the view layer (either react component or old view models), not in repositories

Reply:
@atomrc I changed the implementation to handle all of the logic in calling view model, could you have another look?